ENIGMAsystem/SHELL/Platforms/xlib/XLIBwindow.cpp

```diff
@@ -804,47 +806,107 @@ int window_get_color()
 
 void clipboard_set_text(string text)
 {
-  Atom XA_UTF8 = XInternAtom(disp, "UTF8", 0);
-  Atom XA_CLIPBOARD = XInternAtom(disp, "CLIPBOARD", False);
-  XChangeProperty(disp, RootWindow(disp, 0), XA_CLIPBOARD, XA_UTF8, 8, PropModeReplace, reinterpret_cast<unsigned char*>(const_cast<char*>(text.c_str())), text.length() + 1);
+  //Copy the string into our buffer.
+  if (x11_clipboard) {
+    free(x11_clipboard);
+  }
+  x11_clipboard = strndup(text.c_str(), strlen(text.c_str()));
+
+  //Assert primary ownership
+  XSetSelectionOwner(disp, XA_PRIMARY, win, CurrentTime);
+  Window primOwner = XGetSelectionOwner(disp, XA_PRIMARY);
+  if (primOwner != win) {
+    printf("Warning: XSetSelectionOwner(PRIMARY) failed.\n");
+  }
+
+  //Assert clipboard ownership (this must happen even if we already own the clipboard).
+  XSetSelectionOwner(disp, XA_CLIPBOARD, win, CurrentTime);
+  Window clipOwner = XGetSelectionOwner(disp, XA_CLIPBOARD);
+  if (clipOwner != win) {
+    printf("Warning: XSetSelectionOwner(CLIPBOARD) failed.\n");
+  }
+}
+
+//NOTE: This works as-is with Unicode.
+bool get_clipboard_from_other(Window clipOwner, string* res)
```

Comment:
This should have perhaps been made static and or put into namespace `enigma` away from the user functions.